File: job-matcher/src/scrapers/lever-scraper.js
const BaseScraper = require('./base-scraper');

class LeverScraper extends BaseScraper {
  async scrape() {
    const jobs = [];
    const selectors = this.boardConfig.selectors;
    const maxJobs = this.globalConfig.max_jobs_per_board || 50;
    
    try {
      console.log(`\nScraping Lever board: ${this.boardConfig.name}`);
      await this.page.goto(this.boardConfig.url, { waitUntil: 'networkidle2' });
      
      await this.waitForSelector(selectors.job_list_container, 10000);
      await this.delay(2000);
      
      // Lever typically loads all jobs at once or uses infinite scroll
      // Try scrolling to load all content
      await this.scrollToLoadAll();
      
      const pageJobs = await this.page.evaluate((sel) => {
        const jobCards = document.querySelectorAll(sel.job_card);
        const extracted = [];
        
        jobCards.forEach(card => {
          const titleEl = card.querySelector(sel.title);
          const locationEl = card.querySelector(sel.location);
          const teamEl = card.querySelector(sel.team || '.sort-by-team');
          const linkEl = card.querySelector(sel.apply_link || 'a');
          
          if (titleEl) {
            extracted.push({
              title: titleEl.textContent.trim(),
              location: locationEl ? locationEl.textContent.trim() : 'Not specified',
              team: teamEl ? teamEl.textContent.trim() : '',
              url: linkEl ? linkEl.href : '',
              source: 'Lever'
            });
          }
        });
        
        return extracted;
      }, selectors);
      
      console.log(`  Found ${pageJobs.length} jobs`);
      
      // Get details for each job
      for (const job of pageJobs) {
        if (maxJobs > 0 && jobs.length >= maxJobs) break;
        
        try {
          const jobDetail = await this.getJobDetails(job);
          if (jobDetail) {
            jobs.push(jobDetail);
            console.log(`    ✓ ${job.title}`);
          }
        } catch (e) {
          console.log(`    ✗ Error getting details for ${job.title}: ${e.message}`);
        }
        
        await this.delay(1000);
      }
      
    } catch (error) {
      console.error(`Error scraping Lever board: ${error.message}`);
    }
    
    console.log(`  Total jobs scraped: ${jobs.length}`);
    return jobs;
  }
  
  async scrollToLoadAll() {
    let previousHeight = 0;
    let attempts = 0;
    const maxAttempts = 10;
    
    while (attempts < maxAttempts) {
      const currentHeight = await this.page.evaluate('document.body.scrollHeight');
      if (currentHeight === previousHeight) {
        break;
      }
      
      await this.page.evaluate(() => {
        window.scrollTo(0, document.body.scrollHeight);
      });
      
      await this.delay(1500);
      previousHeight = currentHeight;
      attempts++;
    }
  }
  
  async getJobDetails(job) {
    try {
      // Navigate to job detail page
      if (!job.url) {
        return { ...job, description: '', company: this.boardConfig.name };
      }
      
      await this.page.goto(job.url, { waitUntil: 'networkidle2' });
      await this.delay(2000);
      
      // Extract description
      const description = await this.page.evaluate(() => {
        // Try multiple selectors for Lever job descriptions
        const selectors = [
          '.section.page-centered',
          '[data-qa="job-description"]',
          '.job-description',
          '.description',
          '.posting-content'
        ];
        
        for (const sel of selectors) {
          const el = document.querySelector(sel);
          if (el) {
            return el.textContent.trim();
          }
        }
        
        // Fallback
        return document.body.innerText.substring(0, 3000);
      });
      
      return {
        ...job,
        description: description.substring(0, 3000),
        company: this.boardConfig.name
      };
    } catch (e) {
      return { ...job, description: '', company: this.boardConfig.name };
    }
  }
}

module.exports = LeverScraper;

File: job-matcher/src/scrapers/lever_scraper.py
import json
import logging
import re
from urllib.parse import urljoin

log = logging.getLogger(__name__)


class LeverScraper:
    def __init__(self, board_config):
        self.board_config = board_config

    async def scrape(self, base_scraper):
        url = self.board_config["url"]
        known_urls = set()

        log.info("Scraping Lever board: %s", self.board_config["name"])

        page = await base_scraper.new_page()
        try:
            await base_scraper.goto_with_retry(page, url)

            await page.wait_for_selector(".posting-list, .postings-wrapper", timeout=15000)
            job_cards = await page.query_selector_all(
                "a[href*='/jobs/'], a.posting-title"
            )

            for card in job_cards:
                href = await card.get_attribute("href")
                if not href:
                    continue
                job_url = urljoin(url, href)
                if job_url in known_urls:
                    continue
                known_urls.add(job_url)

                title_el = await card.query_selector("h5, h4, .title")
                title = (await title_el.inner_text()).strip() if title_el else ""

                loc_el = await card.query_selector(".location, .commitment, .workplace-type")
                location = (await loc_el.inner_text()).strip() if loc_el else ""

                team_el = await card.query_selector(".team, .department, .category")
                team = (await team_el.inner_text()).strip() if team_el else ""

                job_desc = ""
                posted_date = ""
                try:
                    detail_page = await base_scraper.new_page()
                    try:
                        await base_scraper.goto_with_retry(detail_page, job_url)
                        desc_el = await detail_page.query_selector(
                            ".content, .posting-description, .description"
                        )
                        if desc_el:
                            job_desc = (await desc_el.inner_text()).strip()

                        body_text = await detail_page.inner_text("body")
                        m = re.search(r"Posted\s+(?:on\s+)?(\w+ \d+,?\s*\d{4})", body_text, re.IGNORECASE)
                        if m:
                            posted_date = m.group(1)
                    finally:
                        await detail_page.close()
                except Exception:
                    log.warning("  Could not load detail for %s", title)

                yield {
                    "url": job_url,
                    "title": title,
                    "company": self.board_config.get("name", ""),
                    "location": location or "Not specified",
                    "description": job_desc or "Not available",
                    "source": "Lever",
                    "team": team,
                    "postedDate": posted_date,
                }

        finally:
            await page.close()
